backend/src/apiserver/resource/resource_manager.go

```diff
@@ -1064,19 +1068,28 @@ func (r *ResourceManager) CreateJob(ctx context.Context, job *model.Job) (*model
 	job.PipelineSpec.PipelineId = pipelineVersion.PipelineId
 	job.PipelineSpec.PipelineVersionId = pipelineVersion.UUID
 	job.PipelineSpec.PipelineName = pipelineVersion.Name
+	if manifest == "" {
+		manifest = pipelineVersion.PipelineSpec
+	}
 	// Get manifest from either of the two places:
 	// (1) raw manifest in pipeline_spec
 	// (2) pipeline version in resource_references
 	// And the latter takes priority over the former when the manifest is from pipeline_spec.pipeline_id
 	// workflow/pipeline manifest and pipeline id/version will not exist at the same time, guaranteed by the validation phase
 	var tmpl template.Template
+	// This should only happen when creating a run from an existing pipeline or pipeline version
 	if wfTemplate == nil {
```

Comment:
My general feeling reading the code, logic seems to be right, but the code is hard to follow. 
We check `if manifest == ""` multiple times in different stages, and also check `if wfTemplate == nil` and patch it separately. `manifest` and `wfTemplate` seems to me the same content that serves different interfaces, they should probably be finalized at the same place. For example, I could imaging we read `manifest` from different possible sources, and generate `wfTemplate` at the end. There's probably some room for refactoring to make the logic easier to follow.

Reply:
I agree. There are a few other places that could be improved within the resource manager. I will try to improve it in the later PRs before GA.